torchchat/distributed/generate.py

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+import asyncio
+import atexit
+import importlib.util
+import subprocess
+import threading
+from abc import abstractmethod
+from collections import deque
+from dataclasses import dataclass
+from functools import partial
+from os import environ
+from pathlib import Path
+from typing import List, Optional
+from uuid import uuid4
+
+import torch.multiprocessing as mp
+from torchchat.cli.builder import BuilderArgs, TokenizerArgs
+from torchchat.distributed.dist_run import NAME_TO_DISTRIBUTION_AND_DTYPE
+
+
+def _setup_env(world_size: int, rank: int, target: callable, *args, **kwargs):
+    environ["MASTER_ADDR"] = "localhost"
+    environ["MASTER_PORT"] = "29500"
+    environ["RDZV_BACKEND"] = "c10d"
+    environ["WORLD_SIZE"] = str(world_size)
+    environ["RANK"] = str(rank)
+    environ["LOCALRANK"] = str(rank)
+
+    return target(*args, **kwargs)
+
+
+def _launch_distributed_inference(
+    model_name: str, builder_args: BuilderArgs, tokenizer_args: TokenizerArgs
+) -> tuple[List]:
+    # create programmatic elastic launch
+    print("Launching distributed inference ...")
+
+    num_processes_per_node = builder_args.pp * builder_args.tp
+
+    from torchchat.distributed.dist_run import main
+
+    mp.set_start_method("spawn")
+
+    pipes = []
+    procs = []
+    for rank in range(num_processes_per_node):
+        server_pipe, client_pipe = mp.Pipe(duplex=True)
+        pipes.append(server_pipe)
+        proc = mp.Process(
+            target=partial(_setup_env, num_processes_per_node, rank, main),
+            args=(model_name, builder_args, tokenizer_args, client_pipe),
+        )
+        proc.start()
+
+    for pipe in pipes:
+        response = pipe.recv()
+
+    print(f"Done launching distributed inference on {num_processes_per_node} GPUs.")
```

Comment:
preferable to use logger.info via the singleton logger  rather than print, so that we get line numbers and rank info added to outputs.